io/src/hash.rs

```diff
@@ -0,0 +1,387 @@
+    |us: &mut hash160::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    ripemd160::HashEngine,
+    |us: &mut ripemd160::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    sha1::HashEngine,
+    |us: &mut sha1::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    sha256::HashEngine,
+    |us: &mut sha256::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    sha256d::HashEngine,
+    |us: &mut sha256d::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    sha384::HashEngine,
+    |us: &mut sha384::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    sha512::HashEngine,
+    |us: &mut sha512::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    sha512_256::HashEngine,
+    |us: &mut sha512_256::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    siphash24::HashEngine,
+    |us: &mut siphash24::HashEngine, buf| {
+        hashes::HashEngine::input(us, buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) }
+);
+
+impl_write!(
+    HmacEngine<T>,
+    |us: &mut HmacEngine<T>, buf| {
+        us.input(buf);
+        Ok(buf.len())
+    },
+    |_us| { Ok(()) },
+    T: hashes::GeneralHash
+);
+
+/// Hashes data from a reader.
+///
+/// Adds functionality to a [`hashes::GeneralHash`] type to support hashing data read from a
+/// buffered reader.
+pub trait HashReader: GeneralHash + sealed::Sealed {
+    /// Hashes the entire contents of the `reader`.
+    fn hash_reader<R: crate::BufRead>(reader: &mut R) -> Result<Self, crate::Error>
+    where
+        Self::Engine: Default,
+    {
+        let mut engine = Self::engine(); // This calls `Self::Engine::default()`.
+        loop {
+            let bytes = reader.fill_buf()?;
+
+            let read = bytes.len();
+            // Empty slice means EOF.
+            if read == 0 {
+                break;
+            }
+
+            engine.input(bytes);
+            reader.consume(read);
+        }
+        Ok(Self::from_engine(engine))
+    }
+}
+
+impl HashReader for hash160::Hash {}
+impl HashReader for ripemd160::Hash {}
+impl HashReader for sha1::Hash {}
+impl HashReader for sha256::Hash {}
+impl HashReader for sha256d::Hash {}
+impl HashReader for sha384::Hash {}
+impl HashReader for sha512::Hash {}
+impl HashReader for sha512_256::Hash {}
+// Note this only works if using the default key.
```

Comment:
What the hell is a "default key"? I don't remember HMAC defining any such thing. This looks broken.

Reply:
Oof. We do, in fact, have a `Default` impl for `HmacEngine` which uses an empty string as a key. Sigh.